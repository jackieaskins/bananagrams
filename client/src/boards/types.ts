import { Tile } from '../tiles/types';

export enum Direction {
  ACROSS = 'ACROSS',
  DOWN = 'DOWN',
}
export type BoardPosition = {
  row: number;
  col: number;
};
export enum ValidationStatus {
  NOT_VALIDATED = 'NOT_VALIDATED',
  VALID = 'VALID',
  INVALID = 'INVALID',
}
export type WordInfo = {
  start: BoardPosition;
  validationStatus: ValidationStatus;
};
export type BoardSquare = {
  tile: Tile;
  validationStatus: ValidationStatus;
};
export type Board = Record<string, BoardSquare | null>;

export const getSquareId = ({ row, col }: BoardPosition): string =>
  `${row},${col}`;
