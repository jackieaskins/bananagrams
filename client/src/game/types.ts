import { Board } from '../board/types';
import { Hand } from '../hand/types';
import { Player } from '../player/types';
import { Tile } from '../tile/types';

export type GameStatus = 'NOT_STARTED' | 'STARTING' | 'IN_PROGRESS' | 'ENDING';
export type Snapshot = {
  players: Player[];
  hands: Record<string, Hand>;
  boards: Record<string, Board>;
};
export type GameInfo = {
  gameId: string;
  gameName: string;
  status: GameStatus;
  countdown: number;
  bunch: Tile[];
  players: Player[];
  hands: Record<string, Hand>;
  boards: Record<string, Board>;
  previousSnapshot: Snapshot | null;
};

export type GameLocationState = { gameInfo: GameInfo };
