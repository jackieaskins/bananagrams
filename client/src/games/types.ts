import { Player } from '../players/types';
import { Tile, TileItem } from '../tiles/types';
import { BoardLocation } from '../boards/types';

export type GameInfo = {
  gameId: string;
  gameName: string;
  isInProgress: boolean;
  bunch: Tile[];
  players: Player[];
  previousSnapshot: Player[] | null;
};

export type GameState = {
  gameInfo: GameInfo;
  handleDump: (tileItem: TileItem) => void;
  handleMoveTileFromBoardToHand: (boardLocation: BoardLocation | null) => void;
  handleMoveTileFromHandToBoard: (
    tileId: string,
    boardLocation: BoardLocation
  ) => void;
  handleMoveTileOnBoard: (
    fromLocation: BoardLocation,
    toLocation: BoardLocation
  ) => void;
  handlePeel: () => void;
  isInGame: boolean;
  walkthroughEnabled: boolean;
};

export type GameLocationState = Partial<GameState>;
