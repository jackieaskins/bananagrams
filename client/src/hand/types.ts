import { Tile } from '../tile/types';

export type Hand = Tile[];
