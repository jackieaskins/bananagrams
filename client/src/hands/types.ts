import { Tile } from '../tiles/types';

export type Hand = Record<string, Tile>;
