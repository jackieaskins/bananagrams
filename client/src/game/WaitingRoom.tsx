import { Card, Col, Row, Space, Typography } from 'antd';
import { useEffect } from 'react';
import { useParams } from 'react-router-dom';

import PlayerTable from '../player/PlayerTable';
import PreviewCarousel from '../preview/PreviewCarousel';
import {
  useGameName,
  usePreviousSnapshot,
  useResetCurrentBoard,
} from './stateHooks';

const { Text } = Typography;
const WaitingRoom = (): JSX.Element => {
  const { gameId } = useParams<{ gameId: string }>();
  const resetCurrentBoard = useResetCurrentBoard();
  const gameName = useGameName();
  const previousSnapshot = usePreviousSnapshot();
  const joinUrl = `${window.location.origin}?gameId=${gameId}&tabKey=joinGame`;

  useEffect(() => {
    resetCurrentBoard();
  }, [resetCurrentBoard]);

  return (
    <Space
      direction="vertical"
      css={{ marginTop: '25px', width: '100%' }}
      size="large"
    >
      <div css={{ textAlign: 'center' }}>
        <h1>{gameName}</h1>
        <Text type="secondary">Invite others to game: </Text>
        <br />
        <Text code copyable type="secondary">
          {joinUrl}
        </Text>
      </div>

      <Row css={{ justifyContent: 'center' }} gutter={12}>
        <Col xs={20} sm={14} md={12} lg={10}>
          <PlayerTable />
        </Col>
        {previousSnapshot && (
          <Col xs={20} sm={14} md={12} lg={10}>
            <Card
              bodyStyle={{ height: '100%' }}
              css={{ backgroundColor: '#fafafa' }}
              size="small"
            >
              <PreviewCarousel
                hands={previousSnapshot.hands}
                boards={previousSnapshot.boards}
                players={previousSnapshot.players}
              />
            </Card>
          </Col>
        )}
      </Row>
    </Space>
  );
};

export default WaitingRoom;
