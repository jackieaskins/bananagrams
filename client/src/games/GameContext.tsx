import React, { createContext, useContext, useEffect, useState } from 'react';
import { useHistory, useLocation, useParams } from 'react-router-dom';

import { useSocket } from '../SocketContext';
import { SetState } from '../state/types';

type GameParams = {
  gameId: string;
};

export type GameLocationState = {
  isInGame?: boolean;
  gameInfo?: GameInfo;
};

export type Player = {
  userId: string;
  username: string;
  isOwner: boolean;
  isReady: boolean;
};

export type GameInfo = {
  gameId: string;
  gameName: string;
  isInProgress: boolean;
  players: Player[];
};

type GameState = {
  gameId: string;
  gameName: string;
  isInGame: boolean;
  isInProgress: boolean;
  setIsInProgress: SetState<boolean>;
  setPlayers: SetState<Player[]>;
  setPlayerReady: (userId: string, isReady: boolean) => void;
  players: Player[];
};

const GameContext = createContext<GameState>({
  gameId: '',
  gameName: '',
  isInGame: false,
  isInProgress: false,
  setIsInProgress: () => undefined,
  setPlayerReady: () => undefined,
  setPlayers: () => undefined,
  players: [],
});

export const GameProvider: React.FC<{}> = ({ children }) => {
  const { socket } = useSocket();
  const { replace } = useHistory();
  const { pathname, state } = useLocation<GameLocationState>();
  const { gameId } = useParams<GameParams>();

  const [gameName] = useState(state?.gameInfo?.gameName ?? '');
  const [isInProgress, setIsInProgress] = useState(
    state?.gameInfo?.isInProgress ?? false
  );
  const [isInGame] = useState(state?.isInGame ?? false);
  const [players, setPlayers] = useState(state?.gameInfo?.players ?? []);

  const setPlayerReady = (userId: string, isReady: boolean): void => {
    setPlayers((prevPlayers) => {
      const index = prevPlayers.findIndex((player) => player.userId === userId);
      return [
        ...prevPlayers.slice(0, index),
        { ...prevPlayers[index], isReady },
        ...prevPlayers.slice(index + 1),
      ];
    });
  };

  useEffect(() => {
    if (isInGame) {
      replace(pathname);

      return (): void => {
        socket.emit('leaveGame', { gameId }, () => undefined);
      };
    }

    return;
  }, []);

  useEffect(() => {
    socket.on('playerJoined', (player: Player) => {
      setPlayers((prevPlayers) => [...prevPlayers, player]);
    });

    socket.on('playerLeft', ({ userId }: Player) =>
      setPlayers((prevPlayers) =>
        prevPlayers.filter((player) => player.userId !== userId)
      )
    );

    socket.on('playerReady', ({ userId }: Player) => {
      console.log(`${userId} is ready`);
      setPlayerReady(userId, true);
    });

    return (): void => {
      socket.off('playerJoined');
      socket.off('playerLeft');
      socket.off('playerReady');
    };
  }, []);

  return (
    <GameContext.Provider
      value={{
        gameId,
        gameName,
        isInGame,
        isInProgress,
        players,
        setIsInProgress,
        setPlayerReady,
        setPlayers,
      }}
    >
      {children}
    </GameContext.Provider>
  );
};

export const useGame = (): GameState => useContext(GameContext);
