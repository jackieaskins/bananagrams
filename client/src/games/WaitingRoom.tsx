import {
  Button,
  HStack,
  Heading,
  Stack,
  Text,
  useClipboard,
  useToast,
} from "@chakra-ui/react";
import { useEffect } from "react";
import { FaRegCopy } from "react-icons/fa6";
import OpponentBoardPreview from "../boards/OpponentBoardPreview";
import PlayerTable from "../players/PlayerTable";
import { useGame } from "./GameContext";

export default function WaitingRoom(): JSX.Element {
  const {
    gameInfo: { gameName, previousSnapshot },
  } = useGame();
  const joinUrl = `${window.location.href}/join`;
  const { onCopy, hasCopied } = useClipboard(joinUrl);
  const toast = useToast();

  useEffect(() => {
    if (hasCopied) {
      toast({ description: "Successfully copied invite link to clipboard" });
    }
  }, [hasCopied, toast]);

  return (
    <Stack marginTop="50px" alignItems="center" spacing={8}>
      <HStack spacing={3} align="center">
        <Heading as="h1" textAlign="center">
          {gameName}
        </Heading>

        <Button
          leftIcon={<FaRegCopy />}
          width="fit-content"
          size="sm"
          onClick={onCopy}
        >
          Copy invite link
        </Button>
      </HStack>

      <Stack
        alignItems={{ base: "center", md: "start" }}
        spacing={8}
        direction={{ base: "column", md: "row" }}
      >
        <PlayerTable />

        {previousSnapshot && (
          <Stack>
            <Text textAlign="center">Here are the boards from that round:</Text>

            <OpponentBoardPreview
              initialPlayerIndex={previousSnapshot.findIndex(
                (player) => player.isTopBanana,
              )}
              players={previousSnapshot}
              includeCurrentPlayer
            />
          </Stack>
        )}
      </Stack>
    </Stack>
  );
}
