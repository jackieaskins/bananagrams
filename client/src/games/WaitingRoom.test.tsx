import { ChakraProvider } from "@chakra-ui/react";
import { screen, waitFor } from "@testing-library/react";
import { gameInfoFixture } from "../fixtures/game";
import { playerFixture } from "../fixtures/player";
import { Player, PlayerStatus } from "../players/types";
import { render } from "../testUtils";
import { useGame } from "./GameContext";
import WaitingRoom from "./WaitingRoom";

const CURRENT_PLAYER_ID = "current-player";

jest.mock("../boards/OpponentBoardPreview", () =>
  jest.fn(() => <div>Preview</div>),
);
jest.mock("../players/PlayerTable", () => jest.fn(() => <table />));

const mockEmit = jest.fn();
jest.mock("../socket/SocketContext", () => ({
  useSocket: () => ({
    socket: {
      id: CURRENT_PLAYER_ID,
      emit: mockEmit,
    },
  }),
}));

const mockUseGame = useGame as jest.Mock;
jest.mock("./GameContext", () => ({
  ...jest.requireActual("./GameContext"),
  useGame: jest.fn(),
}));

function renderRoom(
  players: Player[] = [playerFixture()],
  previousSnapshot?: Player[],
) {
  mockUseGame.mockReturnValue({
    gameInfo: gameInfoFixture({ players, previousSnapshot }),
  });

  return render(
    <ChakraProvider>
      <WaitingRoom />
    </ChakraProvider>,
  );
}

describe("<WaitingRoom />", () => {
  describe("copy invite button", () => {
    async function renderAndClickCopyButton() {
      const { user } = renderRoom();

      await user.click(
        screen.getByRole("button", { name: "Copy invite link" }),
      );
    }

    it("copies the invite link on click", async () => {
      await renderAndClickCopyButton();

      await waitFor(async () => {
        expect(await window.navigator.clipboard.readText()).toBe(
          "http://localhost//join",
        );
      });
    });

    it("shows error toast on error", async () => {
      jest
        .spyOn(window.navigator.clipboard, "writeText")
        .mockRejectedValue(new Error("Can't copy"));

      await renderAndClickCopyButton();

      await waitFor(() => {
        expect(screen.getByRole("status")).toHaveTextContent(
          "Unable to copy invite link, try copying the current page URL directly",
        );
      });
    });

    it("shows success toast on success", async () => {
      await renderAndClickCopyButton();

      await waitFor(() => {
        expect(screen.getByRole("status")).toHaveTextContent(
          "Successfully copied invite link to clipboard",
        );
      });
    });
  });

  describe("start game button", () => {
    function getStartGameButton() {
      return screen.getByRole("button", { name: "Start game" });
    }

    it("is disabled if there are no active players", () => {
      renderRoom([
        playerFixture({ status: PlayerStatus.SPECTATING }),
        playerFixture({ status: PlayerStatus.SPECTATING }),
      ]);

      expect(getStartGameButton()).toBeDisabled();
    });

    it("is disabled if any player is not ready", () => {
      renderRoom([
        playerFixture({ status: PlayerStatus.NOT_READY }),
        playerFixture({ status: PlayerStatus.READY }),
      ]);

      expect(getStartGameButton()).toBeDisabled();
    });

    it("emits split event", async () => {
      const { user } = renderRoom([
        playerFixture({ status: PlayerStatus.SPECTATING }),
        playerFixture({ status: PlayerStatus.SPECTATING }),
        playerFixture({ status: PlayerStatus.READY }),
      ]);

      await user.click(getStartGameButton());

      await waitFor(() => {
        expect(mockEmit).toHaveBeenCalledWith("split", {});
      });
    });
  });

  describe("opponent preview", () => {
    it("does not render opponent preview when no snapshot", () => {
      renderRoom();

      expect(
        screen.queryByText("Here are the boards from that round:"),
      ).not.toBeInTheDocument();
    });

    it("renders opponent preview when snapshot", () => {
      renderRoom([playerFixture()], [playerFixture()]);

      expect(
        screen.getByText("Here are the boards from that round:"),
      ).toBeInTheDocument();
    });
  });
});
