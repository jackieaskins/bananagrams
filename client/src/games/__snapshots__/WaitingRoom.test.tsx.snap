// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WaitingRoom /> does not render board when no snapshot 1`] = `
<Stack
  alignItems="center"
  marginTop="50px"
  spacing={8}
>
  <HStack
    align="center"
    spacing={3}
  >
    <Heading
      as="h1"
      textAlign="center"
    >
      gameName
    </Heading>
    <Button
      leftIcon={<FaRegCopy />}
      onClick={[MockFunction]}
      size="sm"
      width="fit-content"
    >
      Copy invite link
    </Button>
  </HStack>
  <Stack
    alignItems={
      {
        "base": "center",
        "md": "start",
      }
    }
    direction={
      {
        "base": "column",
        "md": "row",
      }
    }
    spacing={8}
  >
    <PlayerTable />
  </Stack>
</Stack>
`;

exports[`<WaitingRoom /> renders game boards when snapshot is present 1`] = `
<Stack
  alignItems="center"
  marginTop="50px"
  spacing={8}
>
  <HStack
    align="center"
    spacing={3}
  >
    <Heading
      as="h1"
      textAlign="center"
    >
      gameName
    </Heading>
    <Button
      leftIcon={<FaRegCopy />}
      onClick={[MockFunction]}
      size="sm"
      width="fit-content"
    >
      Copy invite link
    </Button>
  </HStack>
  <Stack
    alignItems={
      {
        "base": "center",
        "md": "start",
      }
    }
    direction={
      {
        "base": "column",
        "md": "row",
      }
    }
    spacing={8}
  >
    <PlayerTable />
    <Stack>
      <Text
        textAlign="center"
      >
        Here are the boards from that round:
      </Text>
      <OpponentBoardPreview
        includeCurrentPlayer={true}
        initialPlayerIndex={0}
        players={
          [
            {
              "board": [
                [
                  null,
                ],
              ],
              "gamesWon": 0,
              "hand": [],
              "isAdmin": false,
              "isTopBanana": true,
              "status": "NOT_READY",
              "userId": "123",
              "username": "username",
            },
          ]
        }
      />
    </Stack>
  </Stack>
</Stack>
`;
